### ISSUE TITLE
Cant handle switching between X11 and Wayland

### ISSUE THREAD

**Issue by nifker** [jwilm/alacritty#2643]
Which operating system does the issue occur on?
Linux
If on linux, are you using X11 or Wayland?
For this issue both.

Seems like alacritty is not able to be started again on X11 if there is still another Wayland compositor running.
Terminal output:
"wl_drm@36: error 0: authenicate failed
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 71, kind: Other, message: "Protocol error" }', libcore/result.rs:1009:5"

**Comment by chrisduerr** [comment]
I'm not sure if there's any way to reliably determine which display server Alacritty should be launched on with both running. The default behavior is to just go for Wayland whenever possible, so it's likely trying to spawn it there and X11 being active is causing troubles.

Would you mind going into a bit more detail why this is necessary?

**Comment by nifker** [comment]
I thought this could be a nice addition because Im running most applications still on X11 and currently use Wayland as a area for compatibility testing.

**Comment by chrisduerr** [comment]
Running a Wayland compositor inside of X definitely does work. That's usually what I do for testing and Alacritty seems to launch just fine like that. It just launches inside of Wayland and not X11.

How do you think it should be determined where Alacritty will be spawned, assuming that it's even possible?

**Comment by nifker** [comment]
What about checking the XDG_SESSION_TYPE variable?

**Comment by kchibisov** [comment]
`XDG_SESSION_TYPE` is not something very reliable from my point of view. It can be set to `tty` if you've launched your wm  from `tty` (e.g. sway). Of course login mangers (e.g. `sddm`) are responsible for setting such variables ( users should remember to set them manually when launching wms from tty though). However I tend to agree, that using `XDG_SESSION_TYPE` as a "hint" can be a good idea.

It's probably worth testing how `glutin`s window example behaves in this scenario and also worth checking `winit` `v0.19.1`/`master` docs on how to force specific `backend` (As far as I remember it's trying to launch on `wayland` and fallbacks to `X11` if fails).

**Comment by chrisduerr** [comment]
Generally what we currently do (or what winit does to be more specific) is to check WAYLAND_DISPLAY. If this is set to a valid socket, it will be spawned in Wayland.

@nifker Is WAYLAND_DISPLAY set for you when starting Alacritty in X11?

**Comment by nifker** [comment]
No its not the case.

**Comment by kchibisov** [comment]
@nifker just to make it clear for me. You’ve launched some X11 wm with login manager, opened alacritty, then focused some tty, launched some wayland wm, opened alacritty there, then focus your x11 wm and tried to open alacritty, but it fails. Did I understand you correctly here?

**Comment by nifker** [comment]
Thats right.

**Comment by chrisduerr** [comment]
And running `WAYLAND_DISPLAY= alacritty` still does not launch it inside of X11?

**Comment by nifker** [comment]
That seems to work

**Comment by chrisduerr** [comment]
Then your `WAYLAND_DISPLAY` variable is set when launching Alacritty inside of X11, telling Alacritty to launch it using the socket specified in the path that `WAYLAND_DISPLAY` specifies. Thus, Alacritty is doing everything correctly as far as I'm concerned.

**Comment by kchibisov** [comment]
I've took a look at things myself. The behavior is super strange. The `WAYLAND_DISPLAY` variable doesn't seem to be presented when running on `X11`, but it make affects `winit` somehow. I've created the following patch to test this.
```diff
diff --git a/alacritty/src/main.rs b/alacritty/src/main.rs
index 3fa3ff5..fa2f476 100644
--- a/alacritty/src/main.rs
+++ b/alacritty/src/main.rs
@@ -108,6 +108,10 @@ fn main() {
     let persistent_logging = config.persistent_logging();

     // Run alacritty
+    match std::env::var("WAYLAND_DISPLAY") {
+        Ok(val) => println!("WAYLAND_DISPLAY={}", val),
+        Err(_) => println!("NOT PRESENTED"),
+    }
     if let Err(err) = run(config, message_buffer) {
         die!("Alacritty encountered an unrecoverable error:\n\n\t{}\n", Red(err));
     }
```
The output on wayland is fine
```
WAYLAND_DISPLAY=wayland-0
```
but when I'm launching on `X11` its strange
```
NOT PRESENTED
wl_drm@35: error 0: authenicate failed
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 71, kind: Other, message: "Protocol error" }', src/libcore/result.rs:999:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

This issue is presented with `glutin` `v0.21.0` window example, but I wasn't able to repro it on `evlp2` rework and `glutin` master.

**Comment by chrisduerr** [comment]
So this issue is resolved in the latest glutin? Seems like we don't need to worry about it then.

**Comment by kchibisov** [comment]
> , but I wasn't able to repro it on evlp2 rework and glutin master.

> So this issue is resolved in the latest glutin?

Yes, should be added to a list of fixes in #2438 , I guess

**Comment by chrisduerr** [comment]
I'll do that for now, though it might be nice to recheck once #2438 is finalized.

**Comment by kchibisov** [comment]
@chrisduerr Since glutin window example is also fixed, I think if it will be found on #2438 it should be reported directly to `winit`/`glutin` as a regression. I'll recheck for all "strange" `wayland` issues on  #2438 when it'll be ready though.

**Comment by chrisduerr** [comment]
>  it should be reported directly to winit/glutin as a regression

If I understood you correctly, this has been fixed in the latest glutin master? I don't think there is any reason to report a bug that does not exist anymore?

**Comment by kchibisov** [comment]
[1]
> though it might be nice to recheck once #2438 is finalized.

[2]
> Since glutin window example is also fixed, I think if it will be found on #2438 it should be reported directly to winit/glutin as a regression **(in case window example will also became broken at some point)**.

We found out, that the problem was in glutin/winit, you've suggested to recheck #2438 when it'll be ready[1],  I've decided to add a note[2] that in case we'll face this bug again on #2438, we should report this as a regression to `glutin`/`winit` since everything is fine with `glutin`/`winit`/`our evlp2` now.  Sorry, feel a little bit sleepy today.
